File: two_pointers/merge_sorted_arrays.py



def merge(arr1, arr2):
    first_pointer = 0;
    second_pointer = 0;
    merged_array = [];

    while first_pointer < len(arr1) and second_pointer < len(arr2):
        if arr1[first_pointer] < arr2[second_pointer]:
            merged_array.append(arr1[first_pointer])
            first_pointer += 1
        elif arr1[first_pointer] > arr2[second_pointer]:
            merged_array.append(arr2[second_pointer])
            second_pointer += 1
        else:
            merged_array.append(arr1[first_pointer])
            merged_array.append(arr2[second_pointer])
            first_pointer += 1
            second_pointer += 1

    merged_array.extend(arr1[first_pointer:])
    merged_array.extend(arr2[second_pointer:])

    return merged_array




result = merge([2, 4, 6], [1, 3, 5]) # [1, 2, 3, 4, 5, 6]
print(result)

File: two_pointers/merge_sorted_arrays.js
function merge(arr1, arr2) {
  const mergedArray = [];
  let firstPointer = 0;
  let secondPointer = 0;

  while (firstPointer < arr1.length && secondPointer < arr2.length) {
    if (arr1[firstPointer] < arr2[secondPointer]) {
      mergedArray.push(arr1[firstPointer]);
      firstPointer++;
    } else if (arr1[firstPointer] > arr2[secondPointer]) {
      mergedArray.push(arr2[secondPointer]);
      secondPointer++;
    } else {
      mergedArray.push(arr1[firstPointer]);
      mergedArray.push(arr2[secondPointer]);
      firstPointer++;
      secondPointer++;
    }
  }

  return mergedArray.concat(
    arr1.slice(firstPointer),
    arr2.slice(secondPointer),
  );
}

const result = merge([2, 4, 6], [1, 3, 5]) // [1, 2, 3, 4, 5, 6]
const result2 = merge([2], [1, 5, 7]) // [1, 2, 5, 7]
console.log(result)
console.log(result2)